src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker_Generators.cs

```diff
@@ -36,49 +36,71 @@ private sealed partial class RegularCompilationTracker : ICompilationTracker
             Compilation? compilationWithStaleGeneratedTrees,
             CancellationToken cancellationToken)
         {
-            if (creationPolicy.GeneratedDocumentCreationPolicy is GeneratedDocumentCreationPolicy.DoNotCreate)
+            // First try to compute the SG docs in the remote process (if we're the host process), syncing the results
+            // back over to us to ensure that both processes are in total agreement about the SG docs and their
+            // contents.
+            var result = await TryComputeNewGeneratorInfoInRemoteProcessAsync(
+                compilationState, compilationWithoutGeneratedFiles, generatorInfo.Documents, compilationWithStaleGeneratedTrees, creationPolicy.GeneratedDocumentCreationPolicy, cancellationToken).ConfigureAwait(false);
+            if (result.HasValue)
             {
-                // We're frozen.  So we do not want to go through the expensive cost of running generators.  Instead, we
-                // just whatever prior generated docs we have.
-                var generatedSyntaxTrees = await generatorInfo.Documents.States.Values.SelectAsArrayAsync(
-                    static (state, cancellationToken) => state.GetSyntaxTreeAsync(cancellationToken), cancellationToken).ConfigureAwait(false);
+                var updatedCompilationWithGeneratedFiles = result.Value.compilationWithGeneratedFiles;
+                var updatedGeneratedDocuments = result.Value.generatedDocuments;
 
-                var compilationWithGeneratedFiles = compilationWithoutGeneratedFiles.AddSyntaxTrees(generatedSyntaxTrees);
+                // If we only ran the required generators, we need to merge in the older documents for the generators
+                // that didn't run, and add their syntax trees back into the compilation.
+                if (creationPolicy.GeneratedDocumentCreationPolicy == GeneratedDocumentCreationPolicy.CreateOnlyRequired)
+                {
+                    var oldDocumentsToMerge = CalculateOldDocumentsToMerge(generatorInfo.Documents, updatedGeneratedDocuments);
+                    updatedGeneratedDocuments = updatedGeneratedDocuments.AddRange(oldDocumentsToMerge);
+                    updatedCompilationWithGeneratedFiles = updatedCompilationWithGeneratedFiles.AddSyntaxTrees(
+                        await oldDocumentsToMerge.SelectAsArrayAsync(
+                            static (state, cancellationToken) => state.GetSyntaxTreeAsync(cancellationToken), cancellationToken).ConfigureAwait(false));
+                }
```

Comment:
I'm surprised this logic is here rather than being within TryComputeNewGeneratorInfoInRemoteProcessAsync, since it's already looking at the documents and seeing if things can be reused. Specifically, in the case of we had to rerun the Razor generator only, but it didn't produce different outputs, then that's going to missing the optimization in:

            // If we produced just as many documents as before, and none of them required any changes, then we can
            // reuse the prior compilation.

And similar.